torch/optim/optimizer.py

```diff
@@ -200,10 +200,15 @@ def step(self, closure=None):
                     update = self.get_sparse_update(p, **group)
                 else:
                     if weight_decay > 0:
-                        p.grad.add_(weight_decay, p)
+                        l2_grad = weight_decay * p
+                        p.grad.add_(l2_grad)
 
                     update = self.get_update(p, **group)
 
+                    # since gradients are not allowed to change in optimisers
```

Comment:
Adding/Subtracting does not respect the implicit functional contract that `get_update` offers. Moreover, this adds extra steps that, if we ever want to record the gradient in an optimizer, would end up being recorded.